src/Razor/src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs

```diff
@@ -194,6 +196,50 @@ public static void SetFileKind(this RazorCodeDocument document, string fileKind)
             document.Items[typeof(FileKinds)] = fileKind;
         }
 
+        internal static bool TryComputeBaseNamespace(this RazorCodeDocument document, out string baseNamespace)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (FileKinds.IsComponent(document.GetFileKind()))
+            {
+                // If the document or it's imports contains a @namespace directive, we want to use that over the root namespace.
+                var visitor = new ComponentNamespaceDirectiveVisitor();
+                var importSyntaxTrees = document.GetImportSyntaxTrees();
+                if (importSyntaxTrees != null)
+                {
+                    // ImportSyntaxTrees is usually set. Just being defensive.
+                    foreach (var syntaxTree in importSyntaxTrees)
+                    {
+                        visitor.Visit(syntaxTree.Root);
+                    }
+                }
```

Comment:
Yeah, this is a stafeful extension method that shows up everywhere code document is.... I'd really like to see this method defined close to where it's used.

Reply:
I'm kinda Ok with this, but it feels sloppy

Reply:
> I'd really like to see this method defined close to where it's used.

This problem is it is used in two different places that are not really related to each other. I put it here so I didn't have to duplicate this logic. But I still don't like the part where I have to add a special condition for Components. That will go away if I merge the logic for the old and the new namespace directive.

Reply:
Do you plan on doing that?

You've got too sign offs and you're making waffles 😆

Reply:
Yeah I do. I'm investigating the crash that you and Javier reported. I'll get back to this eventually. 😄